### PR TITLE
introduce has_len_all_ranks() to check the length of dataloader across ranks

### LINKED ISSUE #9785: Replace ValueError with warning in `has_len()`
### Motivation

`has_len()` in data.py checks if dataloader has `__len__` method implemented. 
When dataloader has zero length, it raises a ValueError (see code). Recently, we got task from users brought up that when dataset has fewer samples than number of ranks, sampler would split the dataset unevenly, and sometimes 0 batch on certain ranks is inevitable. This ValueError blocked them to have successful PL run. This is also bridging the gap for Detectron2Go to better use Lightning: https://github.com/facebookresearch/d2go

### Pitch

Replace ValueError with warning 
Sync the length of dataloader across ranks instead of checking local dataset. By doing so, we can catch the situation when user unintentionally fit empty dataloader

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ninginthecloud**
## What does this PR do?

Fixes #9785 

This PR has changed following:
1. add new function `has_len_all_ranks()` (I'm not sure if original `has_len()` has been widely used by users in public, so I just add a new util function. Feel free to let me know if I just need to replace `has_len()` directly. ) cc: @tchaton @awaelchli 
2. add new attribute to LightningModuel/LightningDataModule `allow_zero_length_dataloader_with_ddp`
3. check the total length of dataloader across all ranks
4. update tests

test_common.py:test_evaluate[trainer_kwargs2] pytest.param(dict(accelerator="ddp_spawn", gpus=2), marks=RunIf(min_gpus=2))

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/hooks.py [hunk 1]
```diff
@@ -311,9 +311,13 @@ def __init__(self) -> None:
             prepare_data_per_node:
                 If True, each LOCAL_RANK=0 will call prepare data.
                 Otherwise only NODE_RANK=0, LOCAL_RANK=0 will prepare data.
+            allow_zero_length_dataloader_with_ddp:
+                If True, dataloader with zero length within local rank is allowed.
+                Default value is False.
         """
         super().__init__()
         self.prepare_data_per_node: bool = True
+        self.allow_zero_length_dataloader_with_ddp: bool = False
```
Comment (tchaton) [inline pytorch_lightning/core/hooks.py:320]:
```suggestion
        self.allow_zero_length_dataloader_with_multiple_devices: bool = False
```

**Review by tchaton — COMMENTED** [review]
quick question. How does Detectron2 handles un-balanced batches as Lightning would definitely hanged otherwise.

**Comment by awaelchli** [comment]
Hey, just saw this. Sorry for the delay on my response. I won't merge #9764 until 1.6 because of our current feature freeze :)

**Comment by tchaton** [comment]
@ninginthecloud Any chance you can get this PR ready for tomorrow ?

File: CHANGELOG.md [hunk 2]
```diff
@@ -178,12 +178,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed default value of the `max_steps` Trainer argument from `None` to -1 ([#9460](https://github.com/PyTorchLightning/pytorch-lightning/pull/9460))
 - LightningModule now raises an error when calling `log(on_step=False, on_epoch=False)` ([#10227](https://github.com/PyTorchLightning/pytorch-lightning/pull/10227))
 - Quantization aware training observers are now disabled by default during validating/testing/predicting stages ([#8540](https://github.com/PyTorchLightning/pytorch-lightning/pull/8540))
+- Raised `MisconfigurationException` when total length of `dataloader` across ranks is zero, and give warning when total length is non-zero, but only local rank length is zero. ([#9821](https://github.com/PyTorchLightning/pytorch-lightning/pull/9821))
```
Comment (rohitgr7) [inline CHANGELOG.md:181]:
```suggestion
- Raised `MisconfigurationException` when total length of `dataloader` across ranks is zero, and give warning when total length is non-zero, but only local rank length is zero. ([#9827](https://github.com/PyTorchLightning/pytorch-lightning/pull/9827))
```

File: pytorch_lightning/utilities/data.py [hunk 3]
```diff
@@ -75,6 +77,54 @@ def has_len(dataloader: DataLoader) -> bool:
...
+            else:
+                raise MisconfigurationException(
+                    "`Dataloader` within local rank has zero length. Please make sure that it returns at least 1 batch."
+                )
+        else:
+            has_len = True
+    except TypeError:
+        has_len = False
+    except NotImplementedError:  # e.g. raised by torchtext if a batch_size_fn is used
+        has_len = False
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:135]: getting unexpected failures on other tests because of MagicMock.

**Pull request merged**